### ISSUE TITLE
Docker containers for tests

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#549]
Currently, all our tests are run either on developer's box directly (which requires installation of many different packages and maintaining their versions), or at the CI (which relies on individual CI images + configurations, and is not very easy to replicate on developers' side).

@13rac1 has an interesting proposal: leveraging Docker containers to help people run these tests easier. Here's POC: https://github.com/kaitai-io/kaitai_struct/commit/dc959d9079c841c55cdba539dcd6b0d2f42df741

**Comment by GreyCat** [comment]
My comments: I believe it's a very good idea, but it would be much better to have many separate "flows" for different targets (probably many different Docker containers), instead of trying to cram every package for every language into one Docker container (effectively, it was what old CI was doing).

In modern CI, we try to have a more "modular" approach, i.e. VM for a particular target is created one per language. It's faster per VM, possible to run in parallel, etc.

**Comment by KOLANICH** [comment]
>instead of trying to cram every package for every language into one Docker container (effectively, it was what old CI was doing).

I vote for doing both ways.
1. a script setting up testing environment for every language. Target audience - ones who want to create an own docker image.
2. a script creating an image only for a single language. Installs the language, KSC, runtime and contrib build tools runtime. The purpose - CI testing of packages using code compiled from KSC. See my docker images for python.
3. a script creating an image for all languages. Installs ksc, languages and their runtimes. The purpose - CI testing of the compiler itself. No k8s and similar fashionable crap: IMHO per-app self-sufficient packages are just misuse of the technology. Docker images are used in CI because they both make setting up the environment much faster than installing the stuff every time with a script and more efficient than using the environment with everything preinstalled. Per-app Docker images are used because ... it is fashionable, though they don't solve any real problem except providing manufacturers of computer hardware with demand on their hardware and cybercriminals with obsolete vulnerable versions of software installed in these images.

**Comment by 13rac1** [comment]
> instead of trying to cram every package for every language into one Docker container 

> Per-app Docker images are used because ... it is fashionable

There are valid reasons for both sides of this argument. I've built projects both with all-in-one and separate container configurations. My goal is always to reduce the build time. In this case, I'm fairly sure "one container to rule them all" is correct, because we will be waiting on the Travis build queue if there are separate builds for each language. Travis allows only a single build at a time unless [we start paying them](https://travis-ci.com/plans).

> No k8s and similar fashionable crap

I'm not following what k8s has to do with this CI pipeline?

> Per-app Docker images are used because ... it is fashionable

It's far easier to build/maintain/run/monitor one process per container for applications. I don't do it for fashion. :laughing: 

<hr>

My idea:
1. An "out of the box" ability to run the CI on my local machine. Docker is an ideal environment for this purpose. I use Make as a task runner and documentation, because it is ubiquitous and available from every package manager.
2. The same environment should run on the CI system. I'm far more familiar with CircleCI, but I assume Travis is similar.

Recommendations:
1. One Docker container for CI, built by CI process. Perhaps a separate repo? :man_shrugging: The primary goal to build this rarely-changing time-consuming CI container image build out of the regular CI flow.
2. This project, `kaitai_struct`, uses that image (either pinned or `latest`) to run the build/formats/tests CI. Note: On CircleCI one option would require extending `FROM` the build image with an image `COPY`ing the codebase since volume mounts aren't allowed; might be similar on Travis.
3. A developer would run the same command `make test` to pull down the pre-built container.

Current build takes 13 min 28 sec on Travis. My POC, currently doing the majority of time intensive tasks (missing C3, JS and debian packaging), takes just 6 minutes for the 2nd+ build.
```bash
time bash -c "make build;make test"
```

**Comment by KOLANICH** [comment]
>Travis allows only a single build at a time unless we start paying them.

IMHO: GitLab CI is better than travis one.

>I'm not following what k8s has to do with this CI pipeline?

In our case nothing. k8s in CI pipeline is neeeded only when testing an app designed to be multicontainer, but these multicontainer apps are needed extremily rarely.

>An "out of the box" ability to run the CI on my local machine. Docker is an ideal environment for this purpose.

is useful mostly when debugging CI, isn't it? For local debugginh IMHO it is more convenient to use system-installed packages.



>A developer would run the same command make test to pull down the pre-built container.

In fact I don't think that running locally any docker images not built on own machine or other enough trustworthy environment (shared CI is certainly not such an environment) is a good idea. Letter `S` in the word `Docker` stands for `security`.

>One Docker container for CI, built by CI process. Perhaps a separate repo? 🤷‍♂️ The primary goal to build this rarely-changing time-consuming CI container image build out of the regular CI flow.

👍

**Comment by GreyCat** [comment]
@KOLANICH 
> * a script setting up testing environment for every language. Target audience - ones who want to create an own docker image.

They already exist in ci_targets: [prepare-cpp_stl](https://github.com/kaitai-io/ci_targets/blob/master/prepare-cpp_stl), [prepare-csharp](https://github.com/kaitai-io/ci_targets/blob/master/prepare-csharp), [prepare-go](https://github.com/kaitai-io/ci_targets/blob/master/prepare-go), etc.

> * a script creating an image only for a single language. Installs the language, KSC, runtime and contrib build tools runtime. The purpose - CI testing of packages using code compiled from KSC. See my docker images for python.

That's the one I'm talking about.

> * a script creating an image for all languages. Installs ksc, languages and their runtimes. The purpose - CI testing of the compiler itself.

CI testing of "the compiler itself" is a complex multi-step procedure, and it does not even start with "installing ksc", as we'll be "installing sbt" + building ksc from sources.

It's not viable to push everything into one container. For example, we need to test for Linux, Windows and OS X — you can't physically do it in one container. Even for Linux — we need to test with different compilers, versions of base OS, probably different architectures.

> I don't think that folowing that fashion is beneficial. IMHO it is harmful and is a waste of resources.

It's not even a real question of how we do that. We just *can't* go back to monolithic way anymore, we've already crossed that border ages ago.

**Comment by GreyCat** [comment]
@13rac1 
> My goal is always to reduce the build time.

This is per se not a goal. The goal is to make development faster and more sustainable in the long run. If we run builds, like, X% faster, but spend 2-3-10x time maintaining that build infrastructure, it both kills the  purpose of CI and all the motivation to contribute.

> In this case, I'm fairly sure "one container to rule them all" is correct, because we will be waiting on the Travis build queue if there are separate builds for each language. Travis allows only a single build at a time unless we start paying them.

We're not bound to Travis in any way, and, moreover, we're already using Travis, Appveyor and CircleCI. I'm seriously considering adding [Azure pipeline](https://azure.microsoft.com/en-us/services/devops/pipelines/), which allow 10 parallel jobs on a pretty beefy build boxes without any major limitations as well, or GitLab CI, or some [in-browser JavaScript CIs](https://github.com/kaitai-io/kaitai_struct/issues/540).

We currently do have 20 build targets, and it's just scratching the surface. We'll be likely growing at least 2-3x in foreseeable future by doing more variants of the targets, for example:

* JavaScript will likely split into ES3, ES6 and TypeScript,
* for C++ there will be at least C++98/C++11/C++14, may be C++17,
* for C# there will be some kind of older C# (C#2?) + newer C# (likely C#7) + we'll need to run it against Linux/Windows, Mono, traditional .NET frameworks, .NET Standard, .NET Core
* for PHP, there will be "older PHP" and "newer PHP with type safety" variants
* etc, etc.

All that needs rigorous testing.

> My idea:
>
> * An "out of the box" ability to run the CI on my local machine. Docker is an ideal environment for this purpose. I use Make as a task runner and documentation, because it is ubiquitous and available from every package manager.
> * The same environment should run on the CI system. I'm far more familiar with CircleCI, but I assume Travis is similar.

A very good idea, let's do it :) I'm slightly hesitating about Make per se, though — I'm not sure it brings any benefits, and it's not easily available, say, on Windows.

> Current build takes 13 min 28 sec on Travis. My POC, currently doing the majority of time intensive tasks (missing C3, JS and debian packaging), takes just 6 minutes for the 2nd+ build.

Please ignore what you see as a summary time at https://travis-ci.org/kaitai-io/kaitai_struct/jobs/513212381. Current build takes likely:

* ksc + ci_targets stage (~7 minutes, likely can be optimized down to ~4 minutes):
  * 3 min to build ksc
  * 50 sec to build ksc .deb package (likely can be optimized)
  * 47 sec to build ksc JS package (the same)
  * 47 sec to build ksc zip package (the same)
  * 47 sec to run translator tests (the same)
  * ~20-30 sec to build test formats
  * ~30 secs to publish all artifacts
* ci_targets => ci_artifacts stage:
  * [40 minutes](https://travis-ci.org/kaitai-io/ci_targets) at Travis
  * [30 minutes](https://ci.appveyor.com/project/GreyCat/ci-targets) at AppVeyor

It's very easy to make it much faster (likely 2-3-4 minutes for the second stage) by parallelizing it massively.

**Comment by GreyCat** [comment]
@KOLANICH 
> is useful mostly when debugging CI, isn't it? For local debugginh IMHO it is more convenient to use system-installed packages.

It's actually kind of complicated. If we'll be able to run per-language or per-platform parts of CI to validate it using Dockerized environments locally, and it won't take like 10x-100x more time, I'd be super happy to use it, instead of manually keeping my system with all these packages with proper versions installed.

**Comment by KOLANICH** [comment]
>It's not viable to push everything into one container. For example, we need to test for Linux, Windows and OS X — you can't physically do it in one container.

Yes. I mean I  am strictly against using one container per binary - a harmful approach which has become fashionable in last years. It's benefit is horisontal scaling of highload apps, but it has gained popularity in for applications for which it is not beneficial and is even harmful. I am not going to speculate on reasons to do that, but IMHO this approach is unsuitable for most of the projects using it.

>Even for Linux — we need to test with different compilers, versions of base OS, probably different architectures.

Different compilers and runtimes versions can be installed into a single docker image. For different architectures ...
1. the same scripts can be used to build a docker image for all the supported architectures
2. it doesn't require separate images for apps
3. I doubt it is easy to find CI provider supporting something other than x86_64.

BTW, I see you have added Microsoft into your profile. And that you have managed to get the time to work on the project this year. Have you managed to make MS to sponsor the project?

**Comment by GreyCat** [comment]
> Yes. I mean I am strictly against using one container per binary - a harmful approach which has become fashionable in last years.

Yes, yes, we've got your point about your hate of k8s and other "fashionable" approaches. Now let us get back to this discussion, as all that is completely irrelevant to it.

> Different compilers and runtimes versions can be installed into a single docker image.

Yes, my point is that even if something is technically possible to do, it is not a very sustainable approach in long term. Why waste time trying to cram X compilers/runtimes into a single image (and execute them sequentially), while we can work off specialized containers for every particular language, and that would be actually even faster, because they can be execute in parallel.

> 3. I doubt it is easy to find CI provider supporting something other than x86_64.

For example, testing JS in different browsers is essentially "non x86_64". For sake of argument, "x86" is something other than "x86_64" too.

> BTW, I see you have added Microsoft into your profile. And that you have managed to get the time to work on the project this year. Have you managed to make MS to sponsor the project?

I work for Microsoft, it's requirement by the law to disclose that affiliation in GitHub profile. While my day job has some intersections with Kaitai Struct, vast majority of work I do on KS is something I do in my free time, not sponsored and/or endorsed by Microsoft. But, again, let's keep this discussion on topic — I don't see how it's relevant to Docker containers — you have my e-mail if you want to discuss it more.

**Comment by KOLANICH** [comment]
>Why waste time trying to cram X compilers/runtimes into a single image

No reason: it is always easier to take a ready-made image than to create an own one. 

There also will be problems with updating components, it is currently easier to update separate images, because the tools available for Docker for now have no features to split a layer and then add this layer above an existing image.

**Comment by GreyCat** [comment]
Hey folks, I came up with some proof-of-concept solution, check it out:

* Docker images are being declared and built in [docker_images repo](https://github.com/kaitai-io/kaitai_struct_docker_images).
  * To build one locally:
    * Check out that repo.
    * Go to relevant directory (e.g. `src/java/temurin11`)
    * Run `./build`
    * You will get an image named like `kaitai-java-temurin11-linux-x86_64`, which will have all the prerequisites installed for running KS tests for a specific target.
    * If you want to play with that image manually, run `./run-it` in the same directory. It will open an interactive shell (correct version of java/javac or ruby/rspec will be preinstalled there).
  * I will add GitHub Actions-based image building in the CI too, so you don't need to build them locally if you don't want to.
* To run the tests, go to [tests repo](https://github.com/kaitai-io/kaitai_struct_tests) and run `./docker-ci`, e.g. something like this:`./docker-ci -t ruby -i 3.2`
  * Internally, it just runs docker with all the necessary arguments to invoke the right image + mount runtime + mount tests + runs `./ci-$TARGET` in it. There will be some output to the console and `test_out/$TARGET` results generated.

So far I've implemented 3 images/environments:

```
$ ./docker-ci -h
./docker-ci [--shell] --target {TARGET} --implementation {IMPLEMENTATION} [--os {OS}] [--arch {ARCH}]

Available images:
REPOSITORY                           TAG       IMAGE ID       CREATED          SIZE
kaitai-ruby-3.2-linux-x86_64         latest    384455df9608   8 minutes ago    989MB
kaitai-java-temurin17-linux-x86_64   latest    c15f156b2cdb   8 minutes ago    537MB
kaitai-java-temurin11-linux-x86_64   latest    5b2dcbca72d8   10 minutes ago   529MB
```

Any feedback is very welcome :)

**Comment by GreyCat** [comment]
Quick follow-up: here's GitHub Actions building and publishing these images:

* [GH Actions workflow run](https://github.com/kaitai-io/kaitai_struct_docker_images/actions/runs/5425316097/jobs/9865910752)
* [Packages list for our org now](https://github.com/orgs/kaitai-io/packages)
* [Package page](https://github.com/kaitai-io/kaitai_struct_docker_images/pkgs/container/kaitai-java-temurin11-linux-x86_64)

One can pull it straight away without any auth just by doing:

```
docker pull ghcr.io/kaitai-io/kaitai-java-temurin11-linux-x86_64
```
